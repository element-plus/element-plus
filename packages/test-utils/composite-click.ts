import { nextTick } from 'vue'
import triggerEvent from './trigger-event'

import type { DOMWrapper, VueWrapper } from '@vue/test-utils'

const triggerCompositeClick = async <
  T extends VueWrapper<any> | DOMWrapper<Element>
>(
  wrapper: T
) => {
  await wrapper.trigger('mousedown')
  await wrapper.trigger('mouseup')
  await wrapper.trigger('click')
}

export default triggerCompositeClick

export const triggerNativeCompositeClick = async (el: Element) => {
  triggerEvent(el, 'mousedown')
  await nextTick()
  triggerEvent(el, 'mouseup')
  await nextTick()
  triggerEvent(el, 'click')
  return nextTick()
}
