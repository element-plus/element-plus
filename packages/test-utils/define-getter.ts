import { isFunction, isUndefined } from 'lodash-unified'

/**
 *
 * @param obj Object on which to add or modify the property.
 * @param prop The property name
 * @param value The value of `obj[prop]` or a getter
 * @returns A restore function which can reset `obj[prop]`'s value or getter
 */
const defineGetter = (
  obj: Record<string, any>,
  prop: string,
  value: any,
  defaultValue?: any
) => {
  let oldValue = defaultValue
  const { get, configurable } = Object.getOwnPropertyDescriptor(obj, prop) || {}

  if (isUndefined(defaultValue) && isUndefined(get)) {
    try {
      oldValue = obj[prop]
    } catch {
      throw new Error(
        `TypeError: Illegal invocation. Cannot read ${prop} of '${obj}', '${obj}' might be a prototype,  please specify default value instead.`
      )
    }
  }

  const oldGetter = get ?? (() => oldValue)
  const getter = isFunction(value) ? value : () => value

  Object.defineProperty(obj, prop, {
    configurable: true,
    get: getter,
  })

  return () => {
    Object.defineProperty(obj, prop, {
      configurable,
      get: oldGetter,
    })
  }
}

export default defineGetter
