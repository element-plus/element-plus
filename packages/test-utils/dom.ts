export const getCssVariable = (el: HTMLElement, property: string) => {
  return getComputedStyle(el).getPropertyValue(property)
}
