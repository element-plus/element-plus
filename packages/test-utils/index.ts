export * from './mock'
export * from './dom'
export { default as defineGetter } from './define-getter'
export { default as makeScroll } from './make-scroll'
export { default as sleep } from './sleep'
export { default as tick } from './tick'
export { default as triggerEvent } from './trigger-event'
export { default as stylePlugin } from './style-plugin'
export {
  default as triggerCompositeClick,
  triggerNativeCompositeClick,
} from './composite-click'
export { default as makeMount } from './make-mount'
