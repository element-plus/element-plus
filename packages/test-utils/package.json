{
  "name": "@element-plus/test-utils",
  "version": "0.0.5",
  "private": true,
  "license": "MIT",
  "main": "index.ts"
}
