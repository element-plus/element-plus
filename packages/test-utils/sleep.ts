const sleep = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export default sleep
