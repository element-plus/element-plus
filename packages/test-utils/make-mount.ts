import { mount } from '@vue/test-utils'
import { merge } from 'lodash-unified'

const makeMount = <C, O, E>(element: C, defaultOptions: O) => {
  return (props: (E | O) | (E & O) = {} as E) =>
    mount(element, merge({}, defaultOptions, props))
}

interface Options {
  data?: () => {
    [key: string]: unknown
  }
  methods?: {
    [key: string]: (...args: unknown[]) => any
  }
}

export const makeMountFunc = <T extends Record<string, unknown>>(
  defaultOptions: T
) => {
  return (template: string, options: Options) => {
    return mount({
      ...merge({}, defaultOptions, options),
      template,
    })
  }
}

export default makeMount
