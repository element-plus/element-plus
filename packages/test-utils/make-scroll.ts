import sleep from './sleep'

const makeScroll = (
  dom: Element,
  name: 'scrollTop' | 'scrollLeft',
  offset: number
) => {
  const eventTarget = dom === document.documentElement ? window : dom
  dom[name] = offset
  const evt = new CustomEvent('scroll', {
    detail: {
      target: {
        [name]: offset,
      },
    },
  })
  eventTarget.dispatchEvent(evt)
  // must use setTimeout instead of nextTick to wait dom change
  return sleep()
}

export default makeScroll
