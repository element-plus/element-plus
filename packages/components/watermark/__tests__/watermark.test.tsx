import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import Watermark from '../src/watermark.vue'

const AXIOM = 'Rem is the best girl'

describe('Watermark.vue', () => {
  it('create', () => {
    const wrapper = mount(() => (
      <Watermark class="watermark">{AXIOM}</Watermark>
    ))

    expect(wrapper.classes()).toContain('watermark')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('slots', () => {
    const wrapper = mount(() => <Watermark>{AXIOM}</Watermark>)

    expect(wrapper.text()).toContain(AXIOM)
  })
})
