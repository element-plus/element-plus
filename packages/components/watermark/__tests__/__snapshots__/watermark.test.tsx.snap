// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Watermark.vue > create 1`] = `"<div style="position: relative;" class="watermark">Rem is the best girl</div>"`;
